File: android/ai-app-bridge-android/src/main/kotlin/io/github/mobileaidev/aiappbridge/android/capture/CaptureAppend.kt
package io.github.mobileaidev.aiappbridge.android.capture

import org.json.JSONArray
import org.json.JSONObject

internal object CaptureAppend {
    fun appendSanitized(
        store: MobileCaptureStore,
        event: JSONObject,
        targetKey: String,
        runtimeEpoch: String,
    ): AppendReceipt {
        val type = event.getString("type")
        val stream = when (type) {
            "log" -> "logs"
            "network" -> "network"
            "state" -> "state"
            "event" -> "events"
            else -> throw IllegalArgumentException(type)
        }
        return store.append(
            CaptureInput(
                stream = stream,
                targetKey = targetKey,
                runtimeEpoch = runtimeEpoch,
                captureId = event.getLong("id"),
                timestampMs = event.getLong("timestampMs"),
                record = event,
                actionId = event.optString("actionId").takeIf { it.isNotBlank() },
                source = event.getString("source"),
                stateKey = if (stream == "state") {
                    "${event.getString("namespace")}.${event.getString("key")}"
                } else {
                    null
                },
            ),
        )
    }

    fun itemsMatch(oracle: JSONArray, live: JSONArray): Boolean {
        if (oracle.length() != live.length()) return false
        for (index in 0 until oracle.length()) {
            if (!jsonEqual(oracle.get(index), live.get(index))) return false
        }
        return true
    }

    fun valuesMatch(oracle: JSONObject, live: JSONObject): Boolean = jsonEqual(oracle, live)

    private fun jsonEqual(left: Any?, right: Any?): Boolean {
        if (left is JSONObject && right is JSONObject) {
            if (left.length() != right.length()) return false
            val keys = left.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                if (!right.has(key) || !jsonEqual(left.opt(key), right.opt(key))) return false
            }
            return true
        }
        if (left is JSONArray && right is JSONArray) {
            if (left.length() != right.length()) return false
            for (index in 0 until left.length()) {
                if (!jsonEqual(left.get(index), right.get(index))) return false
            }
            return true
        }
        if (left == JSONObject.NULL && right == JSONObject.NULL) return true
        return left == right
    }
}

File: ios/ai-app-bridge-ios/Sources/AiAppBridgeIOS/Capture/CaptureAppend.swift
import CoreFoundation
import Foundation

enum CaptureAppend {
    static func appendSanitized(
        store: MobileCaptureStore,
        event: [String: Any],
        stream: String,
        targetKey: String,
        runtimeEpoch: String,
        completion: @escaping (AppendReceipt) -> Void
    ) {
        switch stream {
        case "logs", "network", "state", "events":
            break
        default:
            completion(invalid("unknown_stream")); return
        }
        let stateKey: String?
        if stream == "state" {
            guard let namespace = event["namespace"] as? String, let key = event["key"] as? String else {
                completion(invalid("invalid_state_key")); return
            }
            stateKey = "\(namespace):\(key)"
        } else {
            stateKey = nil
        }
        guard let captureId = int64(event["id"]), let timestampMs = int64(event["timestampMs"]) else {
            completion(invalid("invalid_capture_identity")); return
        }
        guard let source = event["source"] as? String else {
            completion(invalid("invalid_capture_source")); return
        }
        store.append(
            CaptureInput(
                stream: stream,
                targetKey: targetKey,
                runtimeEpoch: runtimeEpoch,
                captureId: captureId,
                timestampMs: timestampMs,
                record: event,
                actionId: event["actionId"] as? String,
                source: source,
                stateKey: stateKey
            ), completion: completion
        )
    }

    private static func invalid(_ reason: String) -> AppendReceipt {
        AppendReceipt(status: "dropped", accepted: false, committed: false, dropped: true,
                      deduplicated: false, mobileFactId: nil, reason: reason)
    }

    static func response(receipt: AppendReceipt, event: [String: Any]) -> [String: Any] {
        var body: [String: Any] = ["ok": receipt.accepted, "record": event,
         "receipt": ["status": receipt.status, "accepted": receipt.accepted,
                     "committed": receipt.committed, "dropped": receipt.dropped,
                     "mobileFactId": receipt.mobileFactId as Any? ?? NSNull(),
                     "reason": receipt.reason as Any? ?? NSNull()]]
        if !receipt.accepted { body["error"] = receipt.reason ?? "capture_rejected" }
        return body
    }

    private static func int64(_ value: Any?) -> Int64? {
        guard let number = value as? NSNumber, CFGetTypeID(number) != CFBooleanGetTypeID(),
              number.doubleValue >= 0, number.doubleValue <= 9_007_199_254_740_991,
              number.doubleValue == Double(number.int64Value) else { return nil }
        return number.int64Value
    }
}
